src/DiagnosticManager.ts

```diff
@@ -0,0 +1,197 @@
+import type { DiagnosticContext, BsDiagnostic, DiagnosticContextPair } from './interfaces';
+import { URI } from 'vscode-uri';
+import type { AstNode } from './parser/AstNode';
+import type { Scope } from './Scope';
+import { util } from './util';
+import { Cache } from './Cache';
+import { isXmlScope } from './astUtils/reflection';
+import type { BscFile } from './files/BscFile';
+import type { DiagnosticRelatedInformation } from 'vscode-languageserver-protocol';
+
+
+export class DiagnosticManager {
+
+    private multiScopeCache = new Cache<string, { diagnostic: BsDiagnostic; contexts: Set<DiagnosticContext> }>();
+
+    public register(diagnostic: BsDiagnostic, context?: DiagnosticContext) {
```

Comment:
Can we merge this function and the `multiple` function? You could do method overloading like this:
```typescript
public register(diagnostic: BsDiagnostic, context?);
public register(diagnostics: BsDiagnostics[], context);
public register(diagnosticArg: BsDiagnostic|BsDiagnostic[], context){ 
    const diagnostics = Array.isArray(diagnosticArg) ? diagnosticArg: [diagnosticArg];
   for (const diagnostic of diagnostics) { 
    .....stuff
    }
}
```